### ISSUE TITLE
Cannot run `dune clean` on directories/

### ISSUE THREAD

**Issue by hongchangwu** [ocaml/dune#2288]
Most `dune` commands take directories as a positional argument, e.g.

```
dune build <dir>
dune runtest <dir>
dune utop <dir>
```

However I cannot do `dune clean <dir>`, instead I have to do something like `(cd <dir> && dune clean)`. Is there a particular reason why this is the case?

**Comment by diml** [comment]
What would you expect `dune clean <dir>` to do? Cleaning the workspace is an operation that's global to the workspace as there is only one `_build` directory, so it doesn't really make sense for `dune clean` to take a directory argument.

**Comment by hongchangwu** [comment]
I was thinking one reasonable behavior would be to clean all targets reachable from that directory/target?

**Comment by rgrinberg** [comment]
While that makes sense, is it really that useful? Cleaning is not something that should be called at all under a regular workflow. It doesn't make sense to me to spend so much time to improve it.

**Comment by hongchangwu** [comment]
To me it's more about the consistency of the CLI. If I have a repo with multiple package `a, b, c ...`, I can run `dune build a` to build a package at the project root. By symmetry I should be able to run  `dune clean a` to clean it.

Most other modern build tools (e.g. stack for Haskell, cargo for Rust) supports something like that.

**Comment by rgrinberg** [comment]
I don't disagree that it's more aesthetically pleasing. But I fail to see how much `dune clean a` offers over `$ rm -rf _build/a`. Hence the gain from this feature seems quite low and the potential cost in code complexity makes it hard to justify.

If you send a PR that implements this and is reasonable wrt complexity, I'm sure we won't reject it. Otherwise, there's plenty of other things for us to work on.

**Comment by hongchangwu** [comment]
No worries, before spending time on this let's wait and see if others may find it useful.

**Comment by ygrek** [comment]
One reasonable usecase is to see warnings in a directory without rebuilding whole project

**Comment by diml** [comment]
@ygrek is the issue that if you run `dune build` twice in a row, the second time you don't see the warnings? Ideally, Dune would still show the warnings even if doesn't re-run the commands. i.e. it should cache the output of commands across runs.

**Comment by ygrek** [comment]
Yes, exactly. More marginal usecases: changes in external C libraries or tools or something else external to dune.

**Comment by diml** [comment]
Yh that's a good point. Dune does track the external dependencies it knows about, but it certainly can't track them all.
